### PR TITLE
[Kernels] Modular kernel refactor

### PR DESCRIPTION
**Pull request by bnellnm**
## Purpose
Simplify `modular_kernel.py`, mainly the `forward` method of FusedMoEModularKernel.
- Fewer levels of indirection in `forward`
- Simplify `workspace_shape` method by passing in `M` and current/chunked `M` sizes instead of tensors.
- Separate getting the workspace type into a method that defaults to the activation dtype.  Only a few experts can override this.
- Factor out `FusedMoEModularKernel` prepare/finalize code into methods since the async/DBO features have made them more complicated.
- Reduce number of special cases, esp. wrt chunking.
- More robust chunking/reduce/do naive flags in `layer.py`

## Test Plan
Run tests/kernels/moe.

## Test Result
All tests pass.

cc @varun-sundar-rabindranath , @SageMoore , @LucasWilkinson

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 1]
```diff
@@ -597,9 +622,8 @@ def __init__(self):
         self.buffer = None
 
     def get(self, shape: tuple[int, ...], device: torch.device,
-            dtype: torch.dtype):
-        if shape == () or shape is None:
-            return None
+            dtype: torch.dtype) -> torch.Tensor:
+        assert shape != ()
         shape_numel = prod(shape)
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py:645]: Not related to this PR - but below where we are doing a torch.empty if `self.buffer.numel() < shape_numel`, I think it is better to assert that `self.buffer.numel() >= shape_numel` . I think it is easy to make mistakes here and could  cost unintended performance loss.

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:645]: I think we need to do `<` in order to support unit tests since they use various shapes over the run.  So if the existing buffer is too small we'll need to re-allocate.  I don't think this happens much (if at all) in models.  Maybe a warn_once or something would be appropriate?

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 2]
```diff
@@ -1036,3 +956,127 @@ def forward(
...
+                activation=activation,
+                global_num_experts=global_num_experts,
+                expert_map=expert_map,
+                a1q_scale=_chunk_scales(a1q_scale, s, e),
+                a2_scale=_chunk_scales(self.fused_experts.a2_scale, e, e),
+                workspace13=workspace13,
+                workspace2=workspace2,
+                expert_tokens_meta=c_expert_tokens_meta,
+                apply_router_weight_on_input=apply_router_weight_on_input,
+            )
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py]: nit: Can we move the `Invoke Experts` part to a utility function ? that way, like in main https://github.com/vllm-project/vllm/blob/13cdc021732d69e7006cbe2ad564d987e749ce31/vllm/model_executor/layers/fused_moe/modular_kernel.py#L963 , we can detect the empty activations case and skip the `num_chunks=0` case in the other utility functions.  What do you think ?

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:645]: I see. Thanks for the context. This might be a false alarm - I was worried about buffer reuse where we sometimes would use `workspace13` for `output` tensors. But, auditing the few places where this gets called I see that the `.get` method gets called only the respective/correct shapes.  

Also, IIUC this is also a class level object, so i believe during `profile_run` it should grow to its maximum size and during normal execution we shouldn't need to allocate. I am not sure how we'd handle it, in terms of logging, as "growing" the buffer is part of the logic. Spamming `warn` during the profile run could alarm people unnecessarily.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 3]
```diff
@@ -655,256 +679,164 @@ def __init__(
...
+                                            device=a1q.device,
                                             dtype=workspace_dtype)
 
-        assert fused_out is None or fused_out.shape == fused_out_shape, (
-            f"fused_out {fused_out.shape} but expected {fused_out_shape}")
-        if fused_out is None:
-            # reuse workspace13 for the output
+        # Construct the entire output that can then be processed in chunks.
+        if num_chunks == 1 and prod(workspace13_shape) >= prod(
+                fused_out_shape):
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py]: Hey @bnellnm  - on `main` I see that `prod(workspace13_shape) >= prod(fused_out_shape)` is not required for the ` not self.fused_experts.supports_chunking() or num_chunks == 1` case.  Are you seeing this with the refactor ?

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 4]
```diff
@@ -476,34 +484,51 @@ def supports_expert_map(self) -> bool:
...
+        curr_M: int,
         M: int,
         N: int,
         K: int,
         topk: int,
         global_num_experts: int,
         local_num_experts: int,
         expert_tokens_meta: Optional[ExpertTokensMetadata],
-    ) -> tuple[tuple[int, ...], tuple[int, ...], tuple[int, ...], torch.dtype]:
+    ) -> tuple[tuple[int, ...], tuple[int, ...], tuple[int, ...]]:
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py:520]: Looking through some of the implementations, I find the distinction between curr_M and M to be a bit confusing. 
Is it possible to pass a single M ? 
IIUC, the issue is differentiating between the chunked and non-chunked case in computing the `output_shape` vs `workspace13/2_shape` ? assuming that is the only issue,
I think it'd be cleaner to do,
```
   _, _, output_shape = workspace_shapes(M=M)
   ws13, ws2, output_shape=workspace_shapes(M=CHUNK_SIZE)
```
This way experts don't have to worry about the chunking logic. 
do you think this will work ? are there more issues maybe ?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:520]: Maybe I just need to make the names clearer like `M_chunk` and `M_full` or something?  I found calling `workspace_shapes` twice and passing `M`, `a` and the `a1q` tensor to be confusing too.

Let me know what you think after the next update.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 3 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py]: I think this check is still needed depending on whether or not we are using batched or standard format and which experts are being used.  WIth batched format it will always be true but not necessarily for standard.  Also, some experts can have empty workspaces so we'll need to allocate the output in those cases.

I'll add a comment to clarify.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 2 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py]: I've reworked this a bit so that the check is in the main `forward` function.  I've also used `M_chunk` and `M_full` everywhere in order to be more consistent.

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 4 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:520]: If you think it is simpler to call `workspace_shapes` twice with the different `M` sizes, I can go back to that too.  It would be nice to not have to deal with that when implementing that method.

File: tests/kernels/moe/modular_kernel_tools/common.py [hunk 5]
```diff
@@ -209,18 +209,18 @@ def all2all_backend(self):
         info = prepare_finalize_info(self.prepare_finalize_type)
         return info.backend
 
-    def is_valid(self):
+    def is_valid(self) -> tuple[bool, Optional[str]]:
```
Comment (varun-sundar-rabindranath) [inline tests/kernels/moe/modular_kernel_tools/common.py:212]: This is nice! Thanks 🙌

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 6]
```diff
@@ -2207,23 +2199,11 @@ def forward_impl(
...
-            or _use_flashinfer_cutlass_kernels
-        ):
+        if self.use_dp_chunking:
             return self.forward_impl_chunked(hidden_states, router_logits)
 
         do_naive_dispatch_combine: bool = (
-            self.dp_size > 1
-            and not self.moe_parallel_config.use_deepep_ht_kernels
-            and not self.moe_config.use_flashinfer_cutlass_kernels
+            self.dp_size > 1 and self.quant_method.fused_experts is None
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/layer.py]: This is much nicer!  Thanks 🙌 
suggestion : maybe introduce `use_modular_kernel()` in the `quant_method` base class that check if `fused_experts` is None ?

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:520]: Thanks for re-working this !
With the rename, it looks much better than before. But I am leaning towards calling `workspace_shapes` twice as, 
 1. The person implementing that method needs to know about chunking and why it is used - Also with smaller batch sizes `M_chunk` and `M_full` are usually the same and it looks like it'd be easier to use one for another.
 2. With the 2 `workspace_shapes` call, it still is not ideal, but all the logic would be in one place and we can comment it out in that one place. 
 
However, I am good with landing the PR with `M_chunk` and `M_full` and seeing how much friction we get with that 👍  Thanks 🙌

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py]: I see. Yeah, this looks great! Thanks 🙌

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 6 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py]: I added this but I can't use it everywhere due to lint.  The linter complains that uses of the `fused_experts` are potentially None where there used to be an explicit check (which is now pushed into the property).

File: vllm/model_executor/layers/fused_moe/modular_kernel.py (hunk 4 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:520]: Yeah, I think you are right.  I went back to the two `workspace_shapes` calls and single `M` argument.

File: tests/kernels/moe/test_modular_kernel_combinations.py [hunk 7]
```diff
@@ -168,31 +168,92 @@ def is_nyi_config(config: Config) -> bool:
...
 @pytest.mark.parametrize(
-    "combination", product(MK_MULTI_GPU_PREPARE_FINALIZE_TYPES, MK_FUSED_EXPERT_TYPES)
+    "k,n,e,dtype,quant_config,prepare_finalize_type,fused_experts_type,chunk_size,world_size",
+    generate_valid_test_cases(
+        world_size=2, prepare_finalize_types=MK_MULTI_GPU_PREPARE_FINALIZE_TYPES
+    ),
 )
-@pytest.mark.parametrize("fused_moe_chunk_size", FUSED_MOE_CHUNK_SIZEs)
-@pytest.mark.parametrize("world_size", [2])
-@multi_gpu_test(num_gpus=2)
```
Comment (varun-sundar-rabindranath) [inline tests/kernels/moe/test_modular_kernel_combinations.py:181]: is `@multi_gpu_test(num_gpus=2)` not required anymore ?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 8]
```diff
@@ -1235,39 +1239,25 @@ def __init__(
...
-                self.batched_router_logits = torch.zeros(
-                    (moe.max_num_tokens, num_experts),
-                    dtype=moe.in_dtype,
-                    device=torch.cuda.current_device(),
-                )
+            self.batched_hidden_states = torch.zeros(
+                states_shape, dtype=moe.in_dtype, device=torch.cuda.current_device()
+            )
+
+            # Note here we use `num_experts` which is logical expert count
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/layer.py]: nit: looks like cruft ? maybe move this comment to where `logits_shape` is set ?

File: tests/kernels/moe/test_modular_kernel_combinations.py (hunk 7 above)
Comment (bnellnm) [reply tests/kernels/moe/test_modular_kernel_combinations.py:181]: It seems to actually break the tests sometimes.  It forces another level of process spawning which happens to use `fork` instead of `spawn` which causes CUDA to be initialized multiple times.  I've run into this a couple of times now.  I'm not sure why it doesn't happen in 100% of cases.

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 8 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py]: Looks like the comment got moved around due to merges, etc.  I've moved it back to the right spot.

**Review by varun-sundar-rabindranath — APPROVED** [review]
LGTM! Thanks for the much needed refactor 🙌

**Review by mgoin — APPROVED** [review]
Very nice!
